README
Userspace DNS tunnel with support for DoH and DoT
https://www.bamsoftware.com/software/dnstt/
David Fifield <david@bamsoftware.com>
Public domain

dnstt is a DNS tunnel with these features:
 * Works over DNS over HTTPS (DoH) and DNS over TLS (DoT) as well as
   plaintext UDP DNS.
 * Embeds a sequencing and session protocol (KCP/smux), which means that
   the client does not have to wait for a response before sending more
   data, and any lost packets are automatically retransmitted.
 * Encrypts the contents of the tunnel and authenticates the server by
   public key.

dnstt is an application-layer tunnel that runs in userspace. It doesn't
provide a TUN/TAP interface; it only hooks up a local TCP port with a
remote TCP port (like netcat or `ssh -L`) by way of a DNS resolver. It
does not itself provide a SOCKS or HTTP proxy interface, but you can get
the same effect by running a proxy on the tunnel server and having the
tunnel terminate at the proxy.

```
.------.  |            .---------.             .------.
|tunnel|  |            | public  |             |tunnel|
|client|<---DoH/DoT--->|recursive|<--UDP DNS-->|server|
'------'  |c           |resolver |             '------'
   |      |e           '---------'                |
.------.  |n                                   .------.
|local |  |s                                   |remote|
| app  |  |o                                   | app  |
'------'  |r                                   '------'
```


## DNS zone setup

Because the server side of the tunnel acts like an authoritative name
server, you need to own a domain name and set up a subdomain for the
tunnel. Let's say your domain name is example.com and your server's IP
addresses are 203.0.113.2 and 2001:db8::2. Go to your name registrar and
add three new records:

```
A	tns.example.com	points to 203.0.113.2
AAAA	tns.example.com	points to 2001:db8::2
NS	t.example.com	is managed by tns.example.com
```

The labels `tns` and `t` can be anything you want, but the `tns` label
should not be a subdomain of the `t` label (that space is reserved for
the contents of the tunnel), and the `t` label should be short (because
there is limited space available in a DNS message, and the domain name
takes up part of that space).

Now, when a recursive DNS resolver receives a query for a name like
aaaa.t.example.com, it will forward the query to the tunnel server at
203.0.113.2 or 2001:db8::2.


## Tunnel server setup

Compile the server:
```
tunnel-server$ cd dnstt-server
tunnel-server$ go build
```

First you need to generate the server keypair that will be used to
authenticate the server and encrypt the tunnel.
```
tunnel-server$ ./dnstt-server -gen-key -privkey-file server.key -pubkey-file server.pub
privkey written to server.key
pubkey  written to server.pub
```

Run the server. You need to provide an address that will listen for UDP
DNS packets (`:5300`), the private key file (`server.key`), the root of
the DNS zone (`t.example.com`), and a TCP address to which incoming
tunnel streams will be forwarded (`127.0.0.1:8000`).
```
tunnel-server$ ./dnstt-server -udp :5300 -privkey-file server.key t.example.com 127.0.0.1:8000
```

The tunnel server needs to be able to receive packets on an external
port 53. You can have it listen on port 53 directly using `-udp :53`,
but that requires the program to run as root. It is better to run the
program as an ordinary user and have it listen on an unprivileged port
(`:5300` above), and port-forward port 53 to it. On Linux, use these
commands to forward external port 53 to localhost port 5300:
```
tunnel-server$ sudo iptables -I INPUT -p udp --dport 5300 -j ACCEPT
tunnel-server$ sudo iptables -t nat -I PREROUTING -i eth0 -p udp --dport 53 -j REDIRECT --to-ports 5300
tunnel-server$ sudo ip6tables -I INPUT -p udp --dport 5300 -j ACCEPT
tunnel-server$ sudo ip6tables -t nat -I PREROUTING -i eth0 -p udp --dport 53 -j REDIRECT --to-ports 5300
```

You need to also run something for the tunnel server to connect to. It
can be a proxy server or anything else. For testing, you can use an
Ncat listener:
```
tunnel-server$ ncat -l -k -v 127.0.0.1 8000
```


## Tunnel client setup

Compile the client:
```
tunnel-client$ cd dnstt-client
tunnel-client$ go build
```

Copy the server.pub file from the server to the client. You don't need
server.key on the client; leave it on the server.

Choose a public DoH or DoT resolver. There is a list of DoH resolvers
here:
 * https://github.com/curl/curl/wiki/DNS-over-HTTPS#publicly-available-servers

And DoT resolvers here:
 * https://dnsprivacy.org/wiki/display/DP/DNS+Privacy+Public+Resolvers#DNSPrivacyPublicResolvers-DNS-over-TLS%28DoT%29
 * https://dnsencryption.info/imc19-doe.html

To run the tunnel client using DoH, you need to provide the URL of the
DoH resolver (`https://doh.example/dns-query`), the server's public key
files (`server.pub`), the root of the DNS zone (`t.example.com`), and
the local TCP port that will receive connections and forward them
through the tunnel (`127.0.0.1:7000`):
```
tunnel-client$ ./dnstt-client -doh https://doh.example/dns-query -pubkey-file server.pub t.example.com 127.0.0.1:7000
```

For DoT, it's the same, but use the `-dot` option instead:
```
tunnel-client$ ./dnstt-client -dot dot.example:853 -pubkey-file server.pub t.example.com 127.0.0.1:7000
```

Once the tunnel client is running, you can connect to the local end of
the tunnel, type something, and see it appear at the remote end.
```
tunnel-client$ ncat -v 127.0.0.1 7000
```

The client also has a plaintext UDP mode that can work through a
recursive resolver or directly to the tunnel server
(`-udp tns.example.com`), but it does not provide any covertness for the
tunnel and should only be used for testing.


## How to make a proxy

dnstt is only a tunnel; it's up to you what you want to connect to it.
You can make the tunnel work like an ordinary SOCKS or HTTP proxy by
having the tunnel server forward to a standard proxy server. There are
many ways to set it up; here are some examples.


### Ncat HTTP proxy

Ncat has a simple built-in HTTP/HTTPS proxy, good for testing. Be aware
that Ncat's proxy isn't intended for use by untrusted clients; it won't
prevent them from connecting to localhost ports on the tunnel server,
for example.

```
tunnel-server$ ncat -l -k --proxy-type http 127.0.0.1 8000
tunnel-server$ ./dnstt-server -udp :5300 -privkey-file server.key t.example.com 127.0.0.1:8000
```

On the client, have the tunnel client listen on 127.0.0.1:7000, and configure
your applications to use 127.0.0.1:7000 as an HTTP proxy.

```
tunnel-client$ ./dnstt-client -doh https://doh.example/dns-query -pubkey-file server.pub t.example.com 127.0.0.1:7000
tunnel-client$ curl --proxy http://127.0.0.1:7000/ https://wtfismyip.com/text
```


### SSH SOCKS proxy


OpenSSH has a built-in SOCKS proxy, which makes it easy to add a SOCKS
proxy to a server that already has sshd installed.

On the server, make a localhost SSH connection, using the `-D` option to
open a SOCKS listener at port 8000. Then configure the tunnel server to
forward incoming connections to port 8000. Have the tunnel client listen
on its own local port 7000.

```
tunnel-server$ ssh -N -D 127.0.0.1:8000 -o NoHostAuthenticationForLocalhost=yes 127.0.0.1
# Enter the password of the local user on tunnel-server
tunnel-server$ ./dnstt-server -udp :5300 -privkey-file server.key t.example.com 127.0.0.1:8000
```

```
tunnel-client$ ./dnstt-client -doh https://doh.example/dns-query -pubkey-file server.pub t.example.com 127.0.0.1:7000
tunnel-client$ curl --proxy socks5h://127.0.0.1:7000/ https://wtfismyip.com/text
```

The above configuration, by locating the SOCKS client port on the
server, makes a SOCKS proxy that can be used by anyone with access to
the DNS tunnel. Alternatively, you can make an SSH SOCKS proxy for your
own private use, with the SSH connection going through the tunnel and
the SOCKS client port being located at the client.

Let's assume you have the SSH details configured so that you can run
`ssh tunnel-server` on the tunnel client. Make sure `AllowTcpForwarding`
is set to `yes` (the default value) in sshd_config. Run the tunnel
server and have it forward directly to the SSH port.

```
tunnel-server$ ./dnstt-server -udp :5300 -privkey-file server.key t.example.com 127.0.0.1:22
```

Run the tunnel client with the local listening port at 127.0.0.1:7000.
The `HostKeyAlias` ssh option lets you connect to the SSH server as if
it were located at 127.0.0.1:8000. Replace `tunnel-server` with the
hostname or IP address of the SSH server.

```
tunnel-client$ ./dnstt-client -doh https://doh.example/dns-query -pubkey-file server.pub t.example.com 127.0.0.1:8000
tunnel-client$ ssh -N -D 127.0.0.1:7000 -o HostKeyAlias=tunnel-server -p 8000 127.0.0.1
tunnel-client$ curl --proxy socks5h://127.0.0.1:7000/ https://wtfismyip.com/text
```


### Tor bridge

You can run a Tor bridge on the tunnel server and tunnel the connection
to the bridge with dnstt, using dnstt as like a pluggable transport. The
Tor client provides a SOCKS interface that other programs can use. Let's
say your Tor bridge's ORPort is 9001.

```
tunnel-server$ ./dnstt-server -udp :5300 -privkey-file server.key t.example.com 127.0.0.1:9001
```

```
tunnel-client$ ./dnstt-client -doh https://doh.example/dns-query -pubkey-file server.pub t.example.com 127.0.0.1:7000
```

Add a Bridge line to /etc/tor/torrc, or paste it into Tor Browser. You
can get `FINGERPRINT` from /var/lib/tor/fingerprint on the bridge.

```
Bridge 127.0.0.1:7000 FINGERPRINT
```

If you use a system tor, the client SOCKS port will be 127.0.0.1:9050.
If you use Tor Browser, it will be 127.0.0.1:9150.


## Covertness

Support for DoH and DoT is only to make it more difficult for a local
observer to see that a DNS tunnel is being used, not for the overall
security of the connection. There is a separate encryption layer inside
the tunnel that protects the contents of the tunnel from the resolver
itself.

The encryption of DoH or DoT prevents a network observer between the
tunnel client and the resolver from seeing the remote destination of the
tunnel. An observer can see that the tunnel client is connecting to a
resolver, but cannot see where the resolver is forwarding its queries.
An observer can probably infer, based on volume and other traffic
characteristics, that a tunnel is being used, though it cannot tell
where the remote end of the tunnel is, nor what the contents of the
tunnel are. If the tunnel client is not using DoH or DoT but instead UDP
(`-udp` option), then even an observer between the tunnel client and the
resolver can see that a tunnel is being used and where the remote end of
the tunnel is.

An observer between the resolver and the tunnel server (this includes
the resolver itself) can easily tell that a tunnel is being used and
where the remote end of the tunnel is, because there is no DoH or DoT
encryption at that point. This kind of observer still cannot read the
contents of the tunnel, because there is an additional layer of
end-to-end encryption between the tunnel client and the tunnel server.

An observer who watches what leaves the tunnel server will be able to
see anything that the tunnel server forwards to some other host (if the
tunnel server is acting as a proxy, for example), unless that data has
been separately encrypted before being sent through the tunnel.

dnstt-client disguises its TLS fingerprint using uTLS
(https://github.com/refraction-networking/utls). By default, a specific
TLS Client Hello fingerprint is selected randomly from a weighted
distribution. You can control the distribution of fingerprints (or
select a specific single fingerprint) using the `-utls` option. The
syntax of the option's argument is a comma-separated list of fingerprint
names, each optionally preceded by an integer weight and `*`.
```
$ ./dnstt-client -utls '3*Firefox,2*Chrome,1*iOS' ...
$ ./dnstt-client -utls Firefox ...
```
Run `./dnstt-client -help` to see the available fingerprint names and
the default distribution. The special value `none` disables uTLS and
uses the native crypto/tls fingerprint, which is less covert but likely
to be compatible with more servers.


## Encryption and authentication

The tunnel uses a Noise protocol (https://noiseprotocol.org/noise.html)
for end-to-end security between the tunnel client and tunnel server.
This protocol is independent of the DoH or DoT encryption between the
tunnel client and resolver. The specific protocol is Noise_NK_25519_ChaChaPoly_BLAKE2s
(https://noiseprotocol.org/noise.html#protocol-names-and-modifiers).
The NK handshake pattern authenticates the server but not the client.

The Noise layer is sandwiched between two other protocol layers: KCP
(https://github.com/xtaci/kcp-go) which creates a reliable stream on top
of unreliable datagrams, and smux (https://github.com/xtaci/smux) which
provides stream multiplexing and session features. An observer who can
see DNS messages, such as the intermediary resolver, will be able to see
the headers of the KCP layer, but not of the smux layer nor of the
streams that are inside. The model is similar to what you would get with
TLS or SSH over TCP: an observer can see TCP-level ACKs and sequence
numbers, but cannot read the stream data.

```
application data
smux
Noise
KCP
DNS messages
DoH / DoT / UDP DNS
```

When you run `dnstt-server -gen-key`, you can save the private and
public keys to a file using the `-privkey-file` and `-pubkey-file`
options. You can then load the keys later using `-privkey-file` on the
server and `-pubkey-file` on the client. Alternatively, you can deal
with the keys as literal hexadecimal strings rather than files. If you
run `dnstt-server -gen-key` without the `-privkey-file` and
`-pubkey-file` options, it will display the keys rather than save them
to files. You can then use the keys with `-privkey` on the server and
`-pubkey` on the client.
```
$ ./dnstt-server -gen-key
privkey 0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef
pubkey  0000111122223333444455556666777788889999aaaabbbbccccddddeeeeffff
$ ./dnstt-server -udp :5300 -privkey 0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef t.example.com 127.0.0.1:8000
$ ./dnstt-client -dot dot.example:853 -pubkey 0000111122223333444455556666777788889999aaaabbbbccccddddeeeeffff t.example.com 127.0.0.1:7000
```
If you run the server without `-privkey-file` or `-privkey`, it will
generate a temporary keypair and print the public key in the log. But
the key will be different the next time you restart the server, and you
will have to reconfigure clients.


## Payload sizes

In the client, the available space for user data per query depends on
the length of the domain name in use. Shorter domain names leave more
space for user data.

In the server, the available space for user data per response depends on
the maximum UDP payload size. The larger the UDP payload size, the more
space there is for user data. You want to use as large a UDP payload
size as possible, but not larger than what is supported by the resolver
you are using. Values above 1452 may cause IP fragmentation which can
reduce performance. You can control the maximum UDP payload size with
the `-mtu` option on the server. The default is 1232 bytes; this ought
to be supported by most resolvers that understand EDNS(0) (RFC 6891).
For maximum compatibility, set the maximum to 512, but know that doing
so will reduce downstream bandwidth.
```
$ ./dnstt-server -mtu 512 -doh https://doh.example/dns-query -pubkey-file server.pub t.example.com 127.0.0.1:7000
```

The client and server emit an "effective MTU" log line when starting up
that shows how much space is available for user data in each query or
response. For the server, there may be more space available in some
responses and less in others (depending on the size of the corresponding
query); the logged value is the minimum that is guaranteed to be
supported in any response.


CHANGELOG
## v1.20220208.0

-doh and -dot mode use uTLS to camouflage their TLS Client Hello
fingerprint. The fingerprint to use is chosen randomly from a weighted
distribution. You can control this distribution using the new -utls
option. Use "-utls none" to disable uTLS if you encounter TLS errors
with your chosen server. This change means that it is no longer possible
to use a proxy in -doh mode by setting the HTTP_PROXY or HTTPS_PROXY
environment variables; this was an undocumented side effect of using the
Go net/http package with no TLS camouflage.


## v1.20210812.0

Fixed an injection vulnerability in the dnstt-server log message
"NXDOMAIN: not authoritative for %s". This log message contains a
potentially attacker-controlled name. Because DNS labels may contain any
byte value, the log message allowed an attacker to write arbitrary bytes
to the dnstt-server log, with a variety of effects:
* A label containing a newline could break the format of the log, or be
  used to inject false log lines.
* Log output to a terminal could contain terminal escape sequences which
  could, for example, change the color of text, or have worse effects
  with older terminal emulators (https://seclists.org/bugtraq/2003/Feb/315).
* DNS names with a label that contained the dot character (\x2e) would
  be logged in an ambiguous way, with the intra-label dot appearing as a
  label separator.
DNS names are now logged using backslash hex escapes for unusual bytes.
This vulnerability was called to mind by "Injection Attacks Reloaded:
Tunnelling Malicious Payloads over DNS" by Jeitner and Shulman.
https://www.usenix.org/conference/usenixsecurity21/presentation/jeitner

Inhibited some "io: read/write on closed pipe" log messages.


## v1.20210803.0

Performance tuning. Enlarge some buffers and network receive windows to
permit faster downloads in some configurations. Adjusted the dynamic
polling model to work more as intended, and tweaked the server send loop
to continue bundling outgoing packets as long as some are immediately
available.

Reduce idle timeout from 10 minutes to 2 minutes. Terminated sessions
will be reclaimed more quickly.

Documentation updates.


## v0.20210424.0

dnstt was part of a software security audit done by Cure53. The report
found issues of severity levels Low–Medium in dnstt and in one of its
dependencies, a package used for Noise cryptography. This release fixes
the following issues:
 * UCB-02-002: Memory leak in acceptStreams() routine of dnstt server (Low)
 * UCB-02-003: Potential nonce overflow in Noise protocol (Medium)
 * UCB-02-004: Deprecated DH25519 Golang API used by Noise (Low)
 * UCB-02-006: DoS due to unconditional nonce increment (Low)
 * UCB-02-007: DoS due to missing socket timeouts (Low)
Unaddressed in this release are:
 * UCB-02-005: Client ID security considerations & Noise authenticated data (Low)
 * UCB-02-008: Lack of rate limiting in Snowflake and dnstt (Info)
Two other issues in the report, UCB-02-001 and UCB-02-009, do not have
to do with dnstt. For more details and the text of the report, see
https://www.bamsoftware.com/software/dnstt/security.html#cure53-turbotunnel-2021

Added man pages for dnstt-client and dnstt-server.


## v0.20200506.0

Documentation updates.


## v0.20200504.0

Documentation updates and link to web page.
https://www.bamsoftware.com/software/dnstt/


## v0.20200430.0

Send leftover packets (those that were too big to fit in the current DNS
response) to the correct client in the future. Before, such leftover
packets were included in the next response to be sent, regardless of the
client it was intended for.

Fix the loop that packs packets into responses in the server. Before, if
there were not a packet immediately available, the response would always
be empty. Now, a packet has until the response timeout to be included.

Buffer the DoT transport in the client so that length prefixes and DNS
queries happen in the same send.

Don't look for encoded packets in a response that was too short to
contain a ClientID.


## v0.20200426.0

Added the -mtu command-line option to the server.

Documentation fixes.


## v0.20200419.0

Initial public release and announcement.
https://github.com/net4people/bbs/issues/30


dnstt-client/main.go
// dnstt-client is the client end of a DNS tunnel.
//
// Usage:
//     dnstt-client [-doh URL|-dot ADDR|-udp ADDR] -pubkey-file PUBKEYFILE DOMAIN LOCALADDR
//
// Examples:
//     dnstt-client -doh https://resolver.example/dns-query -pubkey-file server.pub t.example.com 127.0.0.1:7000
//     dnstt-client -dot resolver.example:853 -pubkey-file server.pub t.example.com 127.0.0.1:7000
//
// The program supports DNS over HTTPS (DoH), DNS over TLS (DoT), and UDP DNS.
// Use one of these options:
//     -doh https://resolver.example/dns-query
//     -dot resolver.example:853
//     -udp resolver.example:53
//
// You can give the server's public key as a file or as a hex string. Use
// "dnstt-server -gen-key" to get the public key.
//     -pubkey-file server.pub
//     -pubkey 0000111122223333444455556666777788889999aaaabbbbccccddddeeeeffff
//
// DOMAIN is the root of the DNS zone reserved for the tunnel. See README for
// instructions on setting it up.
//
// LOCALADDR is the TCP address that will listen for connections and forward
// them over the tunnel.
//
// In -doh and -dot modes, the program's TLS fingerprint is camouflaged with
// uTLS by default. The specific TLS fingerprint is selected randomly from a
// weighted distribution. You can set your own distribution (or specific single
// fingerprint) using the -utls option. The special value "none" disables uTLS.
//     -utls '3*Firefox,2*Chrome,1*iOS'
//     -utls Firefox
//     -utls none
package main

import (
	"context"
	"crypto/tls"
	"errors"
	"flag"
	"fmt"
	"io"
	"log"
	"net"
	"net/http"
	"os"
	"strings"
	"sync"
	"time"

	utls "github.com/refraction-networking/utls"
	"github.com/xtaci/kcp-go/v5"
	"github.com/xtaci/smux"
	"www.bamsoftware.com/git/dnstt.git/dns"
	"www.bamsoftware.com/git/dnstt.git/noise"
	"www.bamsoftware.com/git/dnstt.git/turbotunnel"
)

// smux streams will be closed after this much time without receiving data.
const idleTimeout = 2 * time.Minute

// dnsNameCapacity returns the number of bytes remaining for encoded data after
// including domain in a DNS name.
func dnsNameCapacity(domain dns.Name) int {
	// Names must be 255 octets or shorter in total length.
	// https://tools.ietf.org/html/rfc1035#section-2.3.4
	capacity := 255
	// Subtract the length of the null terminator.
	capacity -= 1
	for _, label := range domain {
		// Subtract the length of the label and the length octet.
		capacity -= len(label) + 1
	}
	// Each label may be up to 63 bytes long and requires 64 bytes to
	// encode.
	capacity = capacity * 63 / 64
	// Base32 expands every 5 bytes to 8.
	capacity = capacity * 5 / 8
	return capacity
}

// readKeyFromFile reads a key from a named file.
func readKeyFromFile(filename string) ([]byte, error) {
	f, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	return noise.ReadKey(f)
}

// sampleUTLSDistribution parses a weighted uTLS Client Hello ID distribution
// string of the form "3*Firefox,2*Chrome,1*iOS", matches each label to a
// utls.ClientHelloID from utlsClientHelloIDMap, and randomly samples one
// utls.ClientHelloID from the distribution.
func sampleUTLSDistribution(spec string) (*utls.ClientHelloID, error) {
	weights, labels, err := parseWeightedList(spec)
	if err != nil {
		return nil, err
	}
	ids := make([]*utls.ClientHelloID, 0, len(labels))
	for _, label := range labels {
		var id *utls.ClientHelloID
		if label == "none" {
			id = nil
		} else {
			id = utlsLookup(label)
			if id == nil {
				return nil, fmt.Errorf("unknown TLS fingerprint %q", label)
			}
		}
		ids = append(ids, id)
	}
	return ids[sampleWeighted(weights)], nil
}

func handle(local *net.TCPConn, sess *smux.Session, conv uint32) error {
	stream, err := sess.OpenStream()
	if err != nil {
		return fmt.Errorf("session %08x opening stream: %v", conv, err)
	}
	defer func() {
		log.Printf("end stream %08x:%d", conv, stream.ID())
		stream.Close()
	}()
	log.Printf("begin stream %08x:%d", conv, stream.ID())

	var wg sync.WaitGroup
	wg.Add(2)
	go func() {
		defer wg.Done()
		_, err := io.Copy(stream, local)
		if err == io.EOF {
			// smux Stream.Write may return io.EOF.
			err = nil
		}
		if err != nil && !errors.Is(err, io.ErrClosedPipe) {
			log.Printf("stream %08x:%d copy stream←local: %v", conv, stream.ID(), err)
		}
		local.CloseRead()
		stream.Close()
	}()
	go func() {
		defer wg.Done()
		_, err := io.Copy(local, stream)
		if err == io.EOF {
			// smux Stream.WriteTo may return io.EOF.
			err = nil
		}
		if err != nil && !errors.Is(err, io.ErrClosedPipe) {
			log.Printf("stream %08x:%d copy local←stream: %v", conv, stream.ID(), err)
		}
		local.CloseWrite()
	}()
	wg.Wait()

	return err
}

func run(pubkey []byte, domain dns.Name, localAddr *net.TCPAddr, remoteAddr net.Addr, pconn net.PacketConn) error {
	defer pconn.Close()

	ln, err := net.ListenTCP("tcp", localAddr)
	if err != nil {
		return fmt.Errorf("opening local listener: %v", err)
	}
	defer ln.Close()

	mtu := dnsNameCapacity(domain) - 8 - 1 - numPadding - 1 // clientid + padding length prefix + padding + data length prefix
	if mtu < 80 {
		return fmt.Errorf("domain %s leaves only %d bytes for payload", domain, mtu)
	}
	log.Printf("effective MTU %d", mtu)

	// Open a KCP conn on the PacketConn.
	conn, err := kcp.NewConn2(remoteAddr, nil, 0, 0, pconn)
	if err != nil {
		return fmt.Errorf("opening KCP conn: %v", err)
	}
	defer func() {
		log.Printf("end session %08x", conn.GetConv())
		conn.Close()
	}()
	log.Printf("begin session %08x", conn.GetConv())
	// Permit coalescing the payloads of consecutive sends.
	conn.SetStreamMode(true)
	// Disable the dynamic congestion window (limit only by the maximum of
	// local and remote static windows).
	conn.SetNoDelay(
		0, // default nodelay
		0, // default interval
		0, // default resend
		1, // nc=1 => congestion window off
	)
	conn.SetWindowSize(turbotunnel.QueueSize/2, turbotunnel.QueueSize/2)
	if rc := conn.SetMtu(mtu); !rc {
		panic(rc)
	}

	// Put a Noise channel on top of the KCP conn.
	rw, err := noise.NewClient(conn, pubkey)
	if err != nil {
		return err
	}

	// Start a smux session on the Noise channel.
	smuxConfig := smux.DefaultConfig()
	smuxConfig.Version = 2
	smuxConfig.KeepAliveTimeout = idleTimeout
	smuxConfig.MaxStreamBuffer = 1 * 1024 * 1024 // default is 65536
	sess, err := smux.Client(rw, smuxConfig)
	if err != nil {
		return fmt.Errorf("opening smux session: %v", err)
	}
	defer sess.Close()

	for {
		local, err := ln.Accept()
		if err != nil {
			if err, ok := err.(net.Error); ok && err.Temporary() {
				continue
			}
			return err
		}
		go func() {
			defer local.Close()
			err := handle(local.(*net.TCPConn), sess, conn.GetConv())
			if err != nil {
				log.Printf("handle: %v", err)
			}
		}()
	}
}

func main() {
	var dohURL string
	var dotAddr string
	var pubkeyFilename string
	var pubkeyString string
	var udpAddr string
	var utlsDistribution string

	flag.Usage = func() {
		fmt.Fprintf(flag.CommandLine.Output(), `Usage:
  %[1]s [-doh URL|-dot ADDR|-udp ADDR] -pubkey-file PUBKEYFILE DOMAIN LOCALADDR

Examples:
  %[1]s -doh https://resolver.example/dns-query -pubkey-file server.pub t.example.com 127.0.0.1:7000
  %[1]s -dot resolver.example:853 -pubkey-file server.pub t.example.com 127.0.0.1:7000

`, os.Args[0])
		flag.PrintDefaults()
		labels := make([]string, 0, len(utlsClientHelloIDMap))
		labels = append(labels, "none")
		for _, entry := range utlsClientHelloIDMap {
			labels = append(labels, entry.Label)
		}
		fmt.Fprintf(flag.CommandLine.Output(), `
Known TLS fingerprints for -utls are:
`)
		i := 0
		for i < len(labels) {
			var line strings.Builder
			fmt.Fprintf(&line, "  %s", labels[i])
			w := 2 + len(labels[i])
			i++
			for i < len(labels) && w+1+len(labels[i]) <= 72 {
				fmt.Fprintf(&line, " %s", labels[i])
				w += 1 + len(labels[i])
				i++
			}
			fmt.Fprintln(flag.CommandLine.Output(), line.String())
		}
	}
	flag.StringVar(&dohURL, "doh", "", "URL of DoH resolver")
	flag.StringVar(&dotAddr, "dot", "", "address of DoT resolver")
	flag.StringVar(&pubkeyString, "pubkey", "", fmt.Sprintf("server public key (%d hex digits)", noise.KeyLen*2))
	flag.StringVar(&pubkeyFilename, "pubkey-file", "", "read server public key from file")
	flag.StringVar(&udpAddr, "udp", "", "address of UDP DNS resolver")
	flag.StringVar(&utlsDistribution, "utls",
		"3*Firefox_65,1*Firefox_63,1*iOS_12_1",
		"choose TLS fingerprint from weighted distribution")
	flag.Parse()

	log.SetFlags(log.LstdFlags | log.LUTC)

	if flag.NArg() != 2 {
		flag.Usage()
		os.Exit(1)
	}
	domain, err := dns.ParseName(flag.Arg(0))
	if err != nil {
		fmt.Fprintf(os.Stderr, "invalid domain %+q: %v\n", flag.Arg(0), err)
		os.Exit(1)
	}
	localAddr, err := net.ResolveTCPAddr("tcp", flag.Arg(1))
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}

	var pubkey []byte
	if pubkeyFilename != "" && pubkeyString != "" {
		fmt.Fprintf(os.Stderr, "only one of -pubkey and -pubkey-file may be used\n")
		os.Exit(1)
	} else if pubkeyFilename != "" {
		var err error
		pubkey, err = readKeyFromFile(pubkeyFilename)
		if err != nil {
			fmt.Fprintf(os.Stderr, "cannot read pubkey from file: %v\n", err)
			os.Exit(1)
		}
	} else if pubkeyString != "" {
		var err error
		pubkey, err = noise.DecodeKey(pubkeyString)
		if err != nil {
			fmt.Fprintf(os.Stderr, "pubkey format error: %v\n", err)
			os.Exit(1)
		}
	}
	if len(pubkey) == 0 {
		fmt.Fprintf(os.Stderr, "the -pubkey or -pubkey-file option is required\n")
		os.Exit(1)
	}

	utlsClientHelloID, err := sampleUTLSDistribution(utlsDistribution)
	if err != nil {
		fmt.Fprintf(os.Stderr, "parsing -utls: %v\n", err)
		os.Exit(1)
	}
	if utlsClientHelloID != nil {
		log.Printf("uTLS fingerprint %s %s", utlsClientHelloID.Client, utlsClientHelloID.Version)
	}

	// Iterate over the remote resolver address options and select one and
	// only one.
	var remoteAddr net.Addr
	var pconn net.PacketConn
	for _, opt := range []struct {
		s string
		f func(string) (net.Addr, net.PacketConn, error)
	}{
		// -doh
		{dohURL, func(s string) (net.Addr, net.PacketConn, error) {
			addr := turbotunnel.DummyAddr{}
			var rt http.RoundTripper
			if utlsClientHelloID == nil {
				transport := http.DefaultTransport.(*http.Transport).Clone()
				// Disable DefaultTransport's default Proxy =
				// ProxyFromEnvironment setting, for conformity
				// with utlsRoundTripper and with DoT mode,
				// which do not take a proxy from the
				// environment.
				transport.Proxy = nil
				rt = transport
			} else {
				rt = NewUTLSRoundTripper(nil, utlsClientHelloID)
			}
			pconn, err := NewHTTPPacketConn(rt, dohURL, 32)
			return addr, pconn, err
		}},
		// -dot
		{dotAddr, func(s string) (net.Addr, net.PacketConn, error) {
			addr := turbotunnel.DummyAddr{}
			var dialTLSContext func(ctx context.Context, network, addr string) (net.Conn, error)
			if utlsClientHelloID == nil {
				dialTLSContext = (&tls.Dialer{}).DialContext
			} else {
				dialTLSContext = func(ctx context.Context, network, addr string) (net.Conn, error) {
					return utlsDialContext(ctx, network, addr, nil, utlsClientHelloID)
				}
			}
			pconn, err := NewTLSPacketConn(dotAddr, dialTLSContext)
			return addr, pconn, err
		}},
		// -udp
		{udpAddr, func(s string) (net.Addr, net.PacketConn, error) {
			addr, err := net.ResolveUDPAddr("udp", s)
			if err != nil {
				return nil, nil, err
			}
			pconn, err := net.ListenUDP("udp", nil)
			return addr, pconn, err
		}},
	} {
		if opt.s == "" {
			continue
		}
		if pconn != nil {
			fmt.Fprintf(os.Stderr, "only one of -doh, -dot, and -udp may be given\n")
			os.Exit(1)
		}
		var err error
		remoteAddr, pconn, err = opt.f(opt.s)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			os.Exit(1)
		}
	}
	if pconn == nil {
		fmt.Fprintf(os.Stderr, "one of -doh, -dot, or -udp is required\n")
		os.Exit(1)
	}

	pconn = NewDNSPacketConn(pconn, remoteAddr, domain)
	err = run(pubkey, domain, localAddr, remoteAddr, pconn)
	if err != nil {
		log.Fatal(err)
	}
}
